File: coding/src/com/algos/core/li10_tree/bt/Bt_L572_SubtreeOfAnotherTree.java
package  com.algos.core.li10_tree.bt;

import  com.algos.core.li30_model.TreeNode;


public class Bt_L572_SubtreeOfAnotherTree {

	public static void main(String[] args) {
		Bt_L572_SubtreeOfAnotherTree obj = new Bt_L572_SubtreeOfAnotherTree();
		
		TreeNode root = new TreeNode(1);
		root.left = new TreeNode(3);
		root.right = new TreeNode(4);
		root.left.left = new TreeNode(2);
        root.left.right = new TreeNode(6);
        root.left.right.left = new TreeNode(5);
		
		System.out.println(obj.isSubtree(root, root));
	}

	// Problem: https://leetcode.com/problems/subtree-of-another-tree
	// Idea: For each node check if same tree. If not check both in left and right subtrees.
	// Same tree will return quickly if roots are not matching. If roots are matching, then only it will continue.
	public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        if(root == null)
            return false;
        
        if(isSameTree(root, subRoot))
            return true;
            
		return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }
    
    public boolean isSameTree(TreeNode root1, TreeNode root2) {
		if(root1 == null || root2 == null)
			return root1 == root2;
		
        return (root1.val == root2.val) && isSameTree(root1.left, root2.left) && isSameTree(root1.right, root2.right);
	}
}

File: dsa/concepts/10_trees/bt/python/bt_l572__subtree_of_another_tree.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# Problem: https://leetcode.com/problems/subtree-of-another-tree
# Idea: For each node in root, check if it's the same as subRoot.
# If not, recursively check left and right subtrees.
# Time: O(n * m) in worst case, where n = size of root, m = size of subRoot
# Space: O(1) ignoring recursion stack

def is_subtree(root, sub_root):
    if not root:
        return False

    if is_same_tree(root, sub_root):
        return True

    return is_subtree(root.left, sub_root) or is_subtree(root.right, sub_root)

def is_same_tree(root1, root2):
    if not root1 or not root2:
        return root1 is root2

    return (root1.val == root2.val and 
            is_same_tree(root1.left, root2.left) and 
            is_same_tree(root1.right, root2.right))


# -------- Test --------
root = TreeNode(1)
root.left = TreeNode(3)
root.right = TreeNode(4)
root.left.left = TreeNode(2)
root.left.right = TreeNode(6)
root.left.right.left = TreeNode(5)

print(is_subtree(root, root))  # Expected: True
